### ISSUE TITLE
enable weak promotion state

### ISSUE THREAD

**Issue by samrere** [cupy/cupy#7630]
### Description

since cupy runs in GPU, it computes faster when using float32 arrays, so it would be better to keep its data type during operaions. If any operation results in a float64 array, then all subsequent arrays would be float64, making the whole program slow but runs without any error, and you have to check dtype of all arrays to pinpoint the operation that turns a float32 array to a float64 array:
```python
import cupy as cp
x=cp.array([1,2,3]).astype('f')
y=x/123456
print(y.dtype)

float64
```
In numpy they offer an variable called `NPY_PROMOTION_STATE` and when setting to weak the dtype of an array is kept ([doc](https://numpy.org/neps/nep-0050-scalar-promotion.html)):
```python
import os
os.environ['NPY_PROMOTION_STATE']='weak'
import numpy as np
x=np.array([1,2,3]).astype('f')
y=x/123456
print(y.dtype)

float32
```
It would be much better to enable such option in cupy as well.

**Comment by kmaehashi** [comment]
Thanks for the request, @samrere. It looks NEP 50 looks is the NumPy 2.0 roadmap, so we are going to adopt the weak promotion rules eventually.

https://hackmd.io/@seberg/Bk5P9wJuj#Adopt-NEP-50
cc/ @seberg

**Comment by leofang** [comment]
@seberg is there anything else left to do here, or can we just close this now?

**Comment by seberg** [comment]
Yeah, nothing to do besides a bug fix here and there maybe, this is the case in v14
